### PR TITLE
machine/mount/fuse: filesystem improvements

### PR DESCRIPTION
**Pull request by ppknap**
This PR reworks readdir fuse method to seek directory position when requested and stored offsets mismatch. It also adds synchronisation filters for some git files that should not be sent to remote.

### CODE REVIEW

File: go/src/koding/klient/machine/index/index.go [hunk 1]
```diff
@@ -64,6 +64,11 @@ func NewIndexFiles(root string) (*Index, error) {
 			return nil
 		}
 
+		// Don't include OSX trashes to scanned files.
+		if info.Name() == ".Trash" && info.Mode().IsDir() && runtime.GOOS == "darwin" {
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/index/index.go]: Would it make sense to turn it into `isIgnored(fi os.FileInfo) bool` and extend for more ignored files, like vim's swap and temporary ones etc.

File: go/src/koding/klient/machine/mount/notify/fuse/handle.go [hunk 2]
```diff
@@ -35,7 +36,8 @@ type FileHandle struct {
 	InodeID fuseops.InodeID
 	File    *os.File
 
-	size *int64
+	size  int64
+	write int64
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/handle.go]: Looks like only 0/1 is ever assigned to write. What about bool instead?

File: go/src/koding/klient/machine/mount/notify/fuse/handle.go [hunk 3]
```diff
@@ -173,44 +187,173 @@ func (fhg *FileHandleGroup) Close() error {
...
+	dh.mu.Unlock()
+
+	return
+}
+
+// ReadDir writes dirents do provided destination slice. It returns the number
+// of read bytes.
+func (dh *DirHandle) ReadDir(offset fuseops.DirOffset, dst []byte) (n int, err error) {
+	dh.mu.Lock()
+	if offset > fuseops.DirOffset(len(dh.stream)) {
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/handle.go]: `if int(offset) > len(dh.stream) {` - a bit shorter

File: go/src/koding/klient/machine/mount/notify/fuse/handle.go [hunk 4]
```diff
@@ -173,44 +187,173 @@ func (fhg *FileHandleGroup) Close() error {
...
+	ds = make([]fuseutil.Dirent, 0, n.ChildN()+2)
+
+	var offset fuseops.DirOffset = 1
+	n.Children(0, func(child *node.Node) {
+		if !child.Exist() {
+			return
+		}
+
+		ds = append(ds, fuseutil.Dirent{
+			Offset: offset,
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/mount/notify/fuse/handle.go]: Can we use `fuseops.DirOffset(len(ds)) + 1` here instead?

File: go/src/koding/klient/machine/index/index.go (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/klient/machine/index/index.go]: Or e.g. read the list of ignored files from .gitignore

Comment (ppknap) [inline go/src/koding/klient/machine/index/index.go]: @rjeczalik I added this as a workaround for failing tests. Will move mount/skipper.go logic to index which will be more generic solution. But didn't want to do this in this PR.

File: go/src/koding/klient/machine/mount/notify/fuse/handle.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/klient/machine/mount/notify/fuse/handle.go]: It's leftover from the idea to have this atomic and without a mutex. Will change this to bool.

**Review by rjeczalik — APPROVED** [review]

**Comment by rjeczalik** [comment]
Rebased & resolved conflict.

**Review by cihangir — APPROVED** [review]
i am mostly unfamiliar with the logic but as code quality LGTM

**Comment by cihangir** [comment]
has merge conflicts

**Pull request merged**